tools/pre_ci.py

```
@@ -0,0 +1,340 @@
+            command (list[str]): The command to run as a list of strings.
+            description (str): A brief description of the step for logging.
+            fail_fast (bool): If True, exits the script immediately on failure.
+
+        Returns:
+            bool: True if the command succeeded (exit code 0), False otherwise.
+        """
+        print(f"\n>>> [Step: {description}]")
+        env = os.environ.copy()
+        env["PYTHONIOENCODING"] = "utf-8"
+        try:
+            result = subprocess.run(
```

Ruff (0.15.11)

[error] `subprocess` call: check for execution of untrusted input

(S603)